File: packages/evaluatorq/src/lib/table-display.ts
import fs from "node:fs";

import chalk from "chalk";
import { Effect } from "effect";
import stripAnsi from "strip-ansi";

import type { EvaluatorqResult } from "./types.js";

// Color a percentage value: green >= 80%, yellow >= 50%, red below
function colorRate(rate: number): string {
  if (rate >= 80) return chalk.green(`${rate}%`);
  if (rate >= 50) return chalk.yellow(`${rate}%`);
  return chalk.red(`${rate}%`);
}

// Count evaluator scores that use the pass field
function countPassStats(results: EvaluatorqResult): {
  totalWithPass: number;
  totalPassed: number;
} {
  let totalWithPass = 0;
  let totalPassed = 0;

  for (const result of results) {
    for (const job of result.jobResults ?? []) {
      for (const score of job.evaluatorScores ?? []) {
        if (score.score.pass !== undefined) {
          totalWithPass++;
          if (score.score.pass === true) {
            totalPassed++;
          }
        }
      }
    }
  }

  return { totalWithPass, totalPassed };
}

// Truncate string to fit within maxWidth (accounting for ANSI codes)
function truncate(str: string, maxWidth: number): string {
  const plainStr = stripAnsi(str);
  if (plainStr.length <= maxWidth) return str;

  // For strings with ANSI codes, we need to be more careful
  if (str !== plainStr) {
    // Simple approach: just use plain string and re-apply basic coloring if needed
    return `${plainStr.substring(0, maxWidth - 3)}...`;
  }

  return `${str.substring(0, maxWidth - 3)}...`;
}

// Pad string to width (accounting for ANSI codes)
function pad(
  str: string,
  width: number,
  align: "left" | "right" = "left",
): string {
  const actualLength = stripAnsi(str).length;
  const padding = Math.max(0, width - actualLength);

  if (align === "left") {
    return str + " ".repeat(padding);
  } else {
    return " ".repeat(padding) + str;
  }
}

// Get terminal width with fallback
function getTerminalWidth(): number {
  return process.stdout.columns || 80;
}

// Calculate responsive column widths based on terminal width
function calculateResponsiveWidths(
  minWidths: number[],
  terminalWidth: number,
): number[] {
  const numColumns = minWidths.length;
  const separatorWidth = (numColumns - 1) * 3 + 4; // " │ " between columns and "│ " + " │" at edges
  const availableWidth = terminalWidth - separatorWidth;
  const totalMinWidth = minWidths.reduce((sum, w) => sum + w, 0);

  // If terminal is too narrow, just use minimum widths
  if (availableWidth <= totalMinWidth) {
    return minWidths;
  }

  // Calculate extra space and distribute proportionally
  const extraSpace = availableWidth - totalMinWidth;
  return minWidths.map((minWidth) => {
    const proportion = minWidth / totalMinWidth;
    return Math.floor(minWidth + extraSpace * proportion);
  });
}

// Simple table renderer with multi-line support
function renderDetailedTable(
  headers: string[],
  rows: Array<{ main: string[]; details?: string[] }>,
  colWidths: number[],
): string {
  const lines: string[] = [];

  // Calculate total width for detail rows
  // Total width = sum of column widths + (number of columns - 1) * 3 (for " │ " separators) + 4 (for "│ " and " │")
  const totalContentWidth =
    colWidths.reduce((sum, w) => sum + w, 0) + (colWidths.length - 1) * 3;

  // Top border
  const topBorder = `┌${colWidths.map((w) => "─".repeat(w + 2)).join("┬")}┐`;
  lines.push(topBorder);

  // Headers
  const headerRow = `│ ${headers.map((h, i) => pad(h, colWidths[i])).join(" │ ")} │`;
  lines.push(headerRow);

  // Header separator
  const headerSep = `├${colWidths.map((w) => "─".repeat(w + 2)).join("┼")}┤`;
  lines.push(headerSep);

  // Data rows with details
  rows.forEach((row, rowIndex) => {
    // Main row
    const mainRow = `│ ${row.main
      .map((cell, i) => pad(truncate(cell, colWidths[i]), colWidths[i]))
      .join(" │ ")} │`;
    lines.push(mainRow);

    // Detail rows (if any) - these span all columns
    if (row.details && row.details.length > 0) {
      row.details.forEach((detail) => {
        const detailContent = chalk.dim(`  └─ ${detail}`);
        const paddedDetail = pad(detailContent, totalContentWidth);
        const detailRow = `│ ${paddedDetail} │`;
        lines.push(detailRow);
      });
    }

    // Add separator after each row (except last)
    if (rowIndex < rows.length - 1) {
      const rowSep = `├${colWidths.map((w) => "─".repeat(w + 2)).join("┼")}┤`;
      lines.push(rowSep);
    }
  });

  // Bottom border
  const bottomBorder = `└${colWidths.map((w) => "─".repeat(w + 2)).join("┴")}┘`;
  lines.push(bottomBorder);

  return lines.join("\n");
}

// Create a summary display
function createSummaryDisplay(results: EvaluatorqResult): string {
  const totalDataPoints = results.length;
  const failedDataPoints = results.filter((r) => r.error).length;
  const totalJobs = results.reduce(
    (acc, r) => acc + (r.jobResults?.length || 0),
    0,
  );
  const failedJobs = results.reduce((acc, r) => {
    return acc + (r.jobResults?.filter((j) => j.error).length || 0);
  }, 0);

  const successRate =
    totalJobs > 0
      ? Math.round(((totalJobs - failedJobs) / totalJobs) * 100)
      : 0;

  const { totalWithPass, totalPassed } = countPassStats(results);

  const headers = ["Metric", "Value"];
  const rows: Array<{ main: string[]; details?: string[] }> = [
    { main: ["Total Data Points", chalk.cyan(String(totalDataPoints))] },
    {
      main: [
        "Failed Data Points",
        failedDataPoints > 0
          ? chalk.red(String(failedDataPoints))
          : chalk.green("0"),
      ],
    },
    { main: ["Total Jobs", chalk.cyan(String(totalJobs))] },
    {
      main: [
        "Failed Jobs",
        failedJobs > 0 ? chalk.red(String(failedJobs)) : chalk.green("0"),
      ],
    },
    {
      main: [
        "Success Rate",
        failedJobs === 0 ? chalk.green("100%") : colorRate(successRate),
      ],
    },
  ];

  if (totalWithPass > 0) {
    const passRate = Math.round((totalPassed / totalWithPass) * 100);
    const passLabel = `${passRate}% (${totalPassed}/${totalWithPass})`;
    const passColor =
      passRate === 100 ? chalk.green : passRate >= 80 ? chalk.yellow : chalk.red;
    rows.push({
      main: ["Pass Rate", passColor(passLabel)],
    });
  }

  return renderDetailedTable(headers, rows, [20, 15]);
}

// Calculate averages for evaluator scores across all data points
export function calculateEvaluatorAverages(results: EvaluatorqResult): {
  jobNames: string[];
  evaluatorNames: string[];
  averages: Map<
    string,
    Map<string, { value: string; color: (text: string) => string }>
  >;
} {
  // Collect all unique job names and evaluator names
  const allJobNames = new Set<string>();
  const allEvaluatorNames = new Set<string>();

  // Store all scores per evaluator per job
  const scoresByEvaluatorAndJob = new Map<
    string,
    Map<string, (number | boolean | string | Record<string, unknown>)[]>
  >();

  results.forEach((result) => {
    result.jobResults?.forEach((jobResult) => {
      allJobNames.add(jobResult.jobName);
      jobResult.evaluatorScores?.forEach((score) => {
        allEvaluatorNames.add(score.evaluatorName);

        if (!score.error) {
          if (!scoresByEvaluatorAndJob.has(score.evaluatorName)) {
            scoresByEvaluatorAndJob.set(score.evaluatorName, new Map());
          }
          const jobScores = scoresByEvaluatorAndJob.get(score.evaluatorName);
          if (jobScores && !jobScores.has(jobResult.jobName)) {
            jobScores.set(jobResult.jobName, []);
          }
          jobScores?.get(jobResult.jobName)?.push(score.score.value);
        }
      });
    });
  });

  const jobNames = Array.from(allJobNames);
  const evaluatorNames = Array.from(allEvaluatorNames);

  // Calculate averages
  const averages = new Map<
    string,
    Map<string, { value: string; color: (text: string) => string }>
  >();

  evaluatorNames.forEach((evaluatorName) => {
    const evaluatorAverages = new Map<
      string,
      { value: string; color: (text: string) => string }
    >();

    jobNames.forEach((jobName) => {
      const scores =
        scoresByEvaluatorAndJob.get(evaluatorName)?.get(jobName) || [];

      if (scores.length === 0) {
        evaluatorAverages.set(jobName, { value: "-", color: chalk.gray });
      } else {
        const firstScore = scores[0];

        if (typeof firstScore === "number") {
          // Calculate average for numeric scores
          const numericScores = scores as number[];
          const sum = numericScores.reduce((acc, score) => acc + score, 0);
          const avg = sum / numericScores.length;
          evaluatorAverages.set(jobName, {
            value: avg.toFixed(2),
            color: chalk.yellow,
          });
        } else if (typeof firstScore === "boolean") {
          // Calculate pass rate for boolean scores
          const passCount = scores.filter((score) => score === true).length;
          const passRate = (passCount / scores.length) * 100;
          evaluatorAverages.set(jobName, {
            value: `${passRate.toFixed(1)}%`,
            color:
              passRate === 100
                ? chalk.green
                : passRate >= 50
                  ? chalk.yellow
                  : chalk.red,
          });
        } else if (typeof firstScore === "object" && firstScore !== null) {
          // For structured data (e.g., BERT score, ROUGE_N), show placeholder
          evaluatorAverages.set(jobName, {
            value: "[structured]",
            color: chalk.gray,
          });
        } else {
          // For strings, we ignore as per requirements
          evaluatorAverages.set(jobName, {
            value: "[string]",
            color: chalk.gray,
          });
        }
      }
    });

    averages.set(evaluatorName, evaluatorAverages);
  });

  return { jobNames, evaluatorNames, averages };
}

// Create the main results display with new layout
function createResultsDisplay(results: EvaluatorqResult): string {
  if (results.length === 0) return "";

  const { jobNames, evaluatorNames, averages } =
    calculateEvaluatorAverages(results);

  if (jobNames.length === 0 || evaluatorNames.length === 0) {
    return chalk.yellow("No job results or evaluators found.");
  }

  // Build headers: ["Evaluators"] + job names
  const headers = ["Evaluators", ...jobNames];

  // Calculate column widths: evaluator name column + one per job
  const minColWidths = [20, ...jobNames.map(() => 15)];

  const rows: Array<{ main: string[]; details?: string[] }> = [];

  // Create a row for each evaluator
  evaluatorNames.forEach((evaluatorName) => {
    const row: string[] = [evaluatorName];

    jobNames.forEach((jobName) => {
      const avgData = averages.get(evaluatorName)?.get(jobName);
      if (avgData) {
        row.push(avgData.color(avgData.value));
      } else {
        row.push(chalk.gray("-"));
      }
    });

    rows.push({ main: row });
  });

  const terminalWidth = getTerminalWidth();
  const responsiveWidths = calculateResponsiveWidths(
    minColWidths,
    terminalWidth,
  );
  return renderDetailedTable(headers, rows, responsiveWidths);
}

// Collect and format error messages
function collectErrors(results: EvaluatorqResult): string[] {
  const errors: string[] = [];

  results.forEach((result, idx) => {
    if (result.error) {
      errors.push(`• Data point ${idx + 1}: ${result.error.message}`);
    }

    result.jobResults?.forEach((job) => {
      if (job.error) {
        errors.push(`• Job "${job.jobName}": ${job.error.message}`);
      }

      job.evaluatorScores?.forEach((score) => {
        if (score.error) {
          errors.push(
            `• Evaluator "${score.evaluatorName}": ${score.error.message}`,
          );
        }
      });
    });
  });

  return errors;
}

/**
 * Write detailed evaluation results to GitHub Actions step summary.
 * Only writes when GITHUB_STEP_SUMMARY environment variable is set.
 */
function writeGitHubStepSummary(
  evaluationName: string,
  results: EvaluatorqResult,
): void {
  const summaryPath = process.env.GITHUB_STEP_SUMMARY;
  if (!summaryPath) return;

  const totalDataPoints = results.length;
  const failedDataPoints = results.filter((r) => r.error).length;
  const totalJobs = results.reduce(
    (acc, r) => acc + (r.jobResults?.length || 0),
    0,
  );
  const failedJobs = results.reduce((acc, r) => {
    return acc + (r.jobResults?.filter((j) => j.error).length || 0);
  }, 0);

  const { totalWithPass, totalPassed } = countPassStats(results);
  const passRate =
    totalWithPass > 0 ? Math.round((totalPassed / totalWithPass) * 100) : null;
  const statusEmoji =
    passRate !== null
      ? passRate === 100
        ? "✅"
        : "❌"
      : failedJobs > 0
        ? "❌"
        : "✅";

  let markdown = `### ${statusEmoji} ${evaluationName}\n\n`;

  // Summary stats
  markdown += `| Metric | Value |\n`;
  markdown += `|--------|-------|\n`;
  markdown += `| Data Points | ${totalDataPoints} |\n`;
  markdown += `| Failed Data Points | ${failedDataPoints} |\n`;
  markdown += `| Total Jobs | ${totalJobs} |\n`;
  markdown += `| Failed Jobs | ${failedJobs} |\n`;

  if (passRate !== null) {
    markdown += `| **Pass Rate** | **${passRate}% (${totalPassed}/${totalWithPass})** |\n`;
  }

  markdown += `\n`;

  // Evaluator results table
  const { jobNames, evaluatorNames, averages } =
    calculateEvaluatorAverages(results);

  if (jobNames.length > 0 && evaluatorNames.length > 0) {
    markdown += `<details>\n<summary>Detailed Results</summary>\n\n`;
    markdown += `| Evaluator | ${jobNames.join(" | ")} |\n`;
    markdown += `|-----------|${jobNames.map(() => "-------").join("|")}|\n`;

    evaluatorNames.forEach((evaluatorName) => {
      const row = [evaluatorName];
      jobNames.forEach((jobName) => {
        const avgData = averages.get(evaluatorName)?.get(jobName);
        row.push(avgData?.value || "-");
      });
      markdown += `| ${row.join(" | ")} |\n`;
    });

    markdown += `\n</details>\n\n`;
  }

  // Errors
  const errors = collectErrors(results);
  if (errors.length > 0) {
    markdown += `<details>\n<summary>Errors (${errors.length})</summary>\n\n`;
    markdown += "```\n";
    errors.forEach((error) => {
      markdown += `${error}\n`;
    });
    markdown += "```\n\n</details>\n\n";
  }

  markdown += `---\n\n`;

  try {
    fs.appendFileSync(summaryPath, markdown);
  } catch {
    // Silently fail if we can't write to the summary file
  }
}

export const displayResultsTableEffect = (
  results: EvaluatorqResult,
  evaluationName?: string,
): Effect.Effect<void, never, never> =>
  Effect.sync(() => {
    if (results.length === 0) {
      console.log(chalk.yellow("\nNo results to display.\n"));
      return;
    }

    // Write to GitHub step summary if in CI
    if (evaluationName) {
      writeGitHubStepSummary(evaluationName, results);
    }

    // Clear line and move cursor
    console.log("\n");

    // Title
    console.log(chalk.bold.underline.white("EVALUATION RESULTS"));
    console.log("");

    // Summary
    console.log(chalk.bold.white("Summary:"));
    console.log(createSummaryDisplay(results));
    console.log("");

    // Detailed Results
    console.log(chalk.bold.white("Detailed Results:"));
    console.log(createResultsDisplay(results));
    console.log("");

    // Show errors if any
    const errors = collectErrors(results);
    if (errors.length > 0) {
      console.log(chalk.bold.red("Errors:"));
      errors.forEach((error) => {
        console.log(chalk.red(error));
      });
      console.log("");
    }

    // Show tip
    console.log(chalk.dim("💡 Tip: Use print:false to get raw JSON results."));
    console.log("");
  });

File: packages/evaluatorq-py/src/table_display.py
"""Table display for evaluation results using Rich."""

import shutil
from collections import defaultdict
from typing import Any, Dict, List, Set

from rich.console import Console
from rich.table import Table
from rich.text import Text
from rich.panel import Panel
from rich import box

from .types import EvaluatorqResult


def get_terminal_width() -> int:
    """Get terminal width with fallback"""
    return shutil.get_terminal_size(fallback=(80, 24)).columns


def create_summary_display(results: EvaluatorqResult, console: Console) -> Table:
    """Create a summary table of evaluation results"""
    total_data_points = len(results)
    failed_data_points = sum(1 for r in results if r.error)

    total_jobs = sum(len(r.job_results) if r.job_results else 0 for r in results)
    failed_jobs = sum(
        sum(1 for j in r.job_results if j.error) if r.job_results else 0
        for r in results
    )

    success_rate = (
        round(((total_jobs - failed_jobs) / total_jobs) * 100) if total_jobs > 0 else 0
    )

    # Create summary table with borders
    table = Table(show_header=True, header_style="bold", box=box.ROUNDED)
    table.add_column("Metric", style="white", width=20)
    table.add_column("Value", width=15)

    table.add_row("Total Data Points", Text(str(total_data_points), style="cyan"))

    failed_dp_style = "red" if failed_data_points > 0 else "green"
    table.add_row(
        "Failed Data Points", Text(str(failed_data_points), style=failed_dp_style)
    )

    table.add_row("Total Jobs", Text(str(total_jobs), style="cyan"))

    failed_jobs_style = "red" if failed_jobs > 0 else "green"
    table.add_row("Failed Jobs", Text(str(failed_jobs), style=failed_jobs_style))

    # Success rate with color coding
    if failed_jobs == 0:
        rate_style = "green"
    elif success_rate >= 80:
        rate_style = "green"
    elif success_rate >= 50:
        rate_style = "yellow"
    else:
        rate_style = "red"

    table.add_row("Success Rate", Text(f"{success_rate}%", style=rate_style))

    return table


def calculate_evaluator_averages(results: EvaluatorqResult) -> dict[str, Any]:
    """
    Calculate averages for evaluator scores across all data points.

    Returns:
        Dictionary with job_names, evaluator_names, and averages
    """
    all_job_names: Set[str] = set()
    all_evaluator_names: Set[str] = set()

    # Store all scores per evaluator per job
    scores_by_evaluator_and_job: Dict[str, Dict[str, List[float | bool | str]]] = (
        defaultdict(lambda: defaultdict(list))
    )

    # Collect scores
    for result in results:
        if not result.job_results:
            continue

        for job_result in result.job_results:
            all_job_names.add(job_result.job_name)

            if not job_result.evaluator_scores:
                continue

            for score in job_result.evaluator_scores:
                all_evaluator_names.add(score.evaluator_name)

                if not score.error:
                    scores_by_evaluator_and_job[score.evaluator_name][
                        job_result.job_name
                    ].append(score.score.value)

    job_names = sorted(all_job_names)
    evaluator_names = sorted(all_evaluator_names)

    # Calculate averages
    averages: Dict[str, Dict[str, tuple[str, str]]] = {}

    for evaluator_name in evaluator_names:
        evaluator_averages: Dict[str, tuple[str, str]] = {}

        for job_name in job_names:
            scores = scores_by_evaluator_and_job[evaluator_name].get(job_name, [])

            if not scores:
                evaluator_averages[job_name] = ("-", "dim")
            else:
                first_score = scores[0]

                if isinstance(first_score, (int, float)):
                    # Calculate average for numeric scores
                    avg = sum(float(s) for s in scores) / len(scores)
                    evaluator_averages[job_name] = (f"{avg:.2f}", "yellow")

                elif isinstance(first_score, bool):
                    # Calculate pass rate for boolean scores
                    pass_count = sum(1 for s in scores if s is True)
                    pass_rate = (pass_count / len(scores)) * 100

                    if pass_rate == 100:
                        style = "green"
                    elif pass_rate >= 50:
                        style = "yellow"
                    else:
                        style = "red"

                    evaluator_averages[job_name] = (f"{pass_rate:.1f}%", style)

                else:
                    # For strings, show placeholder
                    evaluator_averages[job_name] = ("[string]", "dim")

        averages[evaluator_name] = evaluator_averages

    return {
        "job_names": job_names,
        "evaluator_names": evaluator_names,
        "averages": averages,
    }


def create_results_display(results: EvaluatorqResult, console: Console) -> Table:
    """Create the main results table showing evaluator averages per job"""
    if not results:
        return Table()

    data = calculate_evaluator_averages(results)
    job_names = data["job_names"]
    evaluator_names = data["evaluator_names"]
    averages = data["averages"]

    if not job_names or not evaluator_names:
        # Return empty table or message
        table = Table(show_header=False)
        table.add_row(Text("No job results or evaluators found.", style="yellow"))
        return table

    # Create table with dynamic columns and borders
    table = Table(show_header=True, header_style="bold", box=box.ROUNDED)

    # Add evaluator column
    table.add_column("Evaluators", style="white", width=20)

    # Add job columns
    for job_name in job_names:
        table.add_column(job_name, style="white", width=15, justify="center")

    # Add rows for each evaluator
    for evaluator_name in evaluator_names:
        row_data = [evaluator_name]

        for job_name in job_names:
            avg_value, avg_style = averages[evaluator_name].get(job_name, ("-", "dim"))
            row_data.append(Text(avg_value, style=avg_style))

        table.add_row(*row_data)

    return table


def collect_errors(results: EvaluatorqResult) -> List[str]:
    """Collect and format error messages from results"""
    errors: List[str] = []

    for idx, result in enumerate(results):
        if result.error:
            errors.append(f"• Data point {idx + 1}: {result.error}")

        if result.job_results:
            for job in result.job_results:
                if job.error:
                    errors.append(f'• Job "{job.job_name}": {job.error}')

                if job.evaluator_scores:
                    for score in job.evaluator_scores:
                        if score.error:
                            errors.append(
                                f'• Evaluator "{score.evaluator_name}": {score.error}'
                            )

    return errors


async def display_results_table(results: EvaluatorqResult):
    """
    Display evaluation results in formatted tables.

    Args:
        results: List of DataPointResult objects to display
    """
    console = Console()

    if not results:
        console.print("\n[yellow]No results to display.[/yellow]\n")
        return

    console.print()

    # Title
    console.print("[bold underline white]EVALUATION RESULTS[/bold underline white]")
    console.print()

    # Summary section
    console.print("[bold white]Summary:[/bold white]")
    summary_table = create_summary_display(results, console)
    console.print(summary_table)
    console.print()

    # Detailed results section
    console.print("[bold white]Detailed Results:[/bold white]")
    results_table = create_results_display(results, console)
    console.print(results_table)
    console.print()

    # Show errors if any
    errors = collect_errors(results)
    if errors:
        console.print("[bold red]Errors:[/bold red]")
        for error in errors:
            console.print(f"[red]{error}[/red]")
        console.print()

    # Show tip
    console.print("[dim]💡 Tip: Use print=False to get raw JSON results.[/dim]")
    console.print()

    # Success message (shown after table, matching TypeScript)
    console.print("[green]✓ Evaluation completed successfully[/green]")
    console.print()
